### ISSUE TITLE
Unix domain path too long when using pipe as transport for macOS on Apple Silicon

### ISSUE THREAD

**Issue by testforstephen** [microsoft/vscode-languageserver-node#1346]
VS Code Java extension is experimenting with using pipes as the transport mechanism between the language client and the Java language server. However, the automatically generated pipe paths for macOS on Silicon (such as `/var/folders/c4/f8v01p_s2sn9_hrx6y18dlxr0000gq/T/vscode-b4e4949e2fde772119c33ebd3ad375f0e8b742abf2.sock`) are too long and exceed the OS domain socket limit. See the `generateRandomPipeName` function in language client, it uses 21 random bytes (which is converted to 42 characters) as suffix, can we reduce the random suffix to a lower value?

https://github.com/microsoft/vscode-languageserver-node/blob/d859bb14d1bcb3923eecaf0ef587e55c48502ccc/jsonrpc/src/node/main.ts#L175-L192

See the original issues on Java extension for more context information.
- https://github.com/redhat-developer/vscode-java/issues/3371
- https://github.com/eclipse-jdtls/eclipse.jdt.ls/issues/2932

// @dbaeumer WDYT?

**Comment by testforstephen** [comment]
The tmp directory `/var/folders/…` (generated by `os.tmpdir()`) is actually a symlink to `/private/var/folders/…`, which has a longer path. This causes the pipe path to exceed the 103 characters limit on macOS. We need to find a way to shorten the pipe path for macOS.

**Comment by dbaeumer** [comment]
I recoded the code a little that generates the pipe name.

```js
function generateRandomPipeName() {
    if (process.platform === 'win32') {
        return `\\\\.\\pipe\\lsp-${(0, crypto_1.randomBytes)(16).toString('hex')}-sock`;
    }
    let randomLength = 32;
    const fixedLength = 'lsp-.sock'.length;
    const tmpDir = fs.realpathSync(XDG_RUNTIME_DIR ?? os.tmpdir());
    const limit = safeIpcPathLengths.get(process.platform);
    if (limit !== undefined) {
        randomLength = Math.min(limit - tmpDir.length - fixedLength, randomLength);
    }
    if (randomLength < 16) {
        throw new Error(`Unable to generate a random pipe name with ${randomLength} characters.`);
    }
    const randomSuffix = (0, crypto_1.randomBytes)(Math.floor(randomLength / 2)).toString('hex');
    return path.join(tmpDir, `lsp-${randomSuffix}.sock`);
}
```

@testforstephen could you patch the JS code in the lib folder to see if this works for you under Mac

**Comment by testforstephen** [comment]
@dbaeumer thanks for the quick patch.

I can confirm the patch works well on macOS M2 machine, which generates a new pipe path in the format of `/private/var/folders/j9/2657q9qj5mgc31phjzjz_g8r0000gq/T/lsp-9b5c0dd5f8a89159ce06a26866a2c94c.sock`.

I also tried it in Windows 10 and WSL (ubuntu 20), the new pipe format works.

**Comment by dbaeumer** [comment]
I will prepare a next release so that you can try to publish something.

**Comment by karthiknadig** [comment]
related: https://github.com/microsoft/vscode-languageserver-node/issues/1351

**Comment by dbaeumer** [comment]
I released 9.0.2-next.1 which contains the new code to compute the pipe name
